src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/tools/runInTerminalTool.ts

```diff
@@ -734,6 +740,48 @@ export class RunInTerminalTool extends Disposable implements IToolImpl {
 		}
 	}
 
+	/**
+	 * Enables persistence of terminal command across sessions.
+	 *
+	 * - `terminalCommandUri` stores the command id. When command detection replays with the same id we can look it up directly.
+	 * - `terminalCommandIndex` captures the position within the command history. On reload, the command ID is regenerated, but the
+	 *   command reappears at the same index, so this acts as a fallback identifier.
+	 */
+	private _setTerminalCommandIdentity(toolSpecificData: IChatTerminalToolInvocationData, instance: ITerminalInstance, commandDetection: ICommandDetectionCapability | undefined, explicitCommandId?: string | undefined): void {
+		if (!commandDetection) {
+			return;
+		}
+
+		const commands = commandDetection.commands;
+		if (!commands || commands.length === 0) {
+			return;
+		}
+
+		let resolvedCommand = explicitCommandId ? commands.find(c => c.id === explicitCommandId) : undefined;
+		let resolvedIndex: number | undefined;
+		if (resolvedCommand) {
+			resolvedIndex = commands.indexOf(resolvedCommand);
+		}
+		if (!resolvedCommand) {
+			resolvedCommand = commands.at(-1);
+			resolvedIndex = commands.length - 1;
+		}
+		if (!resolvedCommand || resolvedIndex === undefined || resolvedIndex < 0) {
+			return;
+		}
+
+		if (!toolSpecificData.terminalCommandUri && resolvedCommand.id) {
+			const params = new URLSearchParams(instance.resource.query);
+			params.set('command', resolvedCommand.id);
+			const commandUri = instance.resource.with({ query: params.toString() || undefined });
+			toolSpecificData.terminalCommandUri = commandUri.toJSON();
+		}
+
+		if (toolSpecificData.terminalCommandIndex === undefined) {
+			toolSpecificData.terminalCommandIndex = resolvedIndex;
+		}
```

Comment:
`toolSpecificData` is meant to be immutable at this point, it's purely for passing from prepare to invoke. The `TerminalChatService` should be used to create this link.